moveos/moveos-verifier/src/verifier.rs

```diff
@@ -1,10 +1,80 @@
-use anyhow::{Error, Result};
-use move_core_types::resolver::MoveResolver;
+use anyhow::{Error, Ok, Result};
+use move_binary_format::file_format::Visibility;
+use move_binary_format::{access::ModuleAccess, CompiledModule};
+use move_core_types::move_resource::MoveStructType;
+use move_core_types::{identifier::Identifier, resolver::MoveResolver};
 use move_vm_runtime::session::{LoadedFunctionInstantiation, Session};
 use move_vm_types::loaded_data::runtime_types::Type;
+use moveos_types::move_types::FunctionId;
+use moveos_types::storage_context::StorageContext;
+use once_cell::sync::Lazy;
 use std::ops::Deref;
 
+pub static INIT_FN_NAME_IDENTIFIER: Lazy<Identifier> =
+    Lazy::new(|| Identifier::new("init").unwrap());
+
+/// The initializer function must have the following properties in order to be executed at publication:
+/// - Name init
+/// - Single parameter of &mut TxContext type
+/// - No return values
+/// - Private
+pub fn verify_init_function<S>(module: &CompiledModule, session: &Session<S>) -> Result<bool>
+where
+    S: MoveResolver,
+{
+    for fdef in &module.function_defs {
+        let fhandle = module.function_handle_at(fdef.function);
+        let fname = module.identifier_at(fhandle.name);
+        if fname == INIT_FN_NAME_IDENTIFIER.clone().as_ident_str() {
+            if Visibility::Private != fdef.visibility {
+                return Err(Error::msg("init function should private".to_string()));
+            } else if fdef.is_entry {
+                return Err(Error::msg(
+                    "init function should not entry function".to_string(),
+                ));
+            } else {
+                let function_id =
+                    FunctionId::new(module.self_id(), INIT_FN_NAME_IDENTIFIER.clone());
+                let loaded_function = session.load_function(
+                    &module.self_id(),
+                    &function_id.function_name,
+                    vec![].as_slice(),
+                )?;
+                let parameters_usize = loaded_function.parameters.len();
+                if parameters_usize != 1 {
+                    return Err(Error::msg(
+                        "init function only should have a parameter with storageContext"
+                            .to_string(),
+                    ));
+                }
+                for ref ty in loaded_function.parameters {
+                    match ty {
+                        Type::Struct(s) | Type::StructInstantiation(s, _) => {
+                            let struct_type = session.get_struct_type(*s).unwrap();
+                            if *struct_type.module.address()
+                                == *moveos_types::addresses::MOVEOS_STD_ADDRESS
+                                && struct_type.module.name()
+                                    == StorageContext::module_identifier().as_ident_str()
+                                && struct_type.name == StorageContext::struct_identifier()
```

Comment:
The checking logic code used here for checking the init function is inconsistent with the logic in the check_init_module function in [metadata.rs](http://metadata.rs/). It should be consistent and abstract a common checking logic.

Reply:
This problem has been resolved